nixos/modules/services/networking/nar-serve.nix

```
@@ -0,0 +1,57 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.nar-serve;
+in
+{
+  meta = {
+    maintainers = [ maintainers.rizary ];
+  };
+  options = {
+    services.nar-serve = {
+      enable = mkEnableOption "serve NAR file content with a webserver";
+
+      port = mkOption {
+        type = types.int;
+        default = 8383;
+        description = ''
+          Port number where nar-serve will listen on.
+        '';
+      };
+
+      nixCacheURL = mkOption {
+        type = types.str;
+        default = "https://cache.nixos.org/";
+        description = ''
+          Binary cache URL to connect to.
+
+          nar-serve also compatible with nix remote url style, such as:
+          - HttpBinaryCacheStore (http://, https://) for binary caches via HTTP or HTTPS
+          - S3BinaryCacheStore (s3://) for binary caches stored in Amazon S3
+          - LegacySSHStore (ssh://) for implement remote builds and nix-copy-closure
+          - SSHStore (ssh-ng://) for arbitrary Nix operations on a remote machine via the same protocol used by nix-daemon
```

nar-serve doesn't support the ssh protocols
```suggestion
          - http://, https:// for binary caches via HTTP or HTTPS
          - s3:// for binary caches stored in Amazon S3
          - gs:// for binary caches stored in Google Cloud Storage
```

---

Does `nar-serve` reimplement the store implementations? Seems better to use commands like `nix cat-nar --store <...>` since then it would support all stores automatically.

---

It's a pure Go implementation at the moment. Having our own implementation will allow to add support for on-disk caching for example.

---

This also allowed us to easily add Google Cloud Storage as a supported backend.